### PR TITLE
feat(table) Add support for date picker in filter section

### LINKED ISSUE #3431: [Table] Add support for datepicker in filters
### What package is this for?

- [x] React
- [ ] Angular

### Summary

Add support to filter by dates using date picker.

**Additional context**
This is how it will look.

Screen Shot 2022-05-06 at 16 39 37

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release?

_NB: The core contributors will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### Available extra resources

What resources do you have to assist this effort?

_Carbon add-ons for Watson IoT is a collaborative system. We encourage teams to
build components and submit them for integration._

### PR DESCRIPTION
**Pull request by mapv04**
Closes #3431

**Summary**
- Add support to filter by dates using date picker instead of typing the date 

**Change List (commits, features, bugs, etc)**

- Add date picker when isDate is true

**Acceptance Test (how to verify the PR)**

- Navigate to the advanced filter and try to filter data using the date picker.
Screen Shot 2022-05-06 at 17 03 26

**Regression Test (how to make sure this PR doesn't break old functionality)**

- If you don't provide the isDate flag then it will use the default behavior.

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx [hunk 1]
```diff
@@ -309,7 +311,39 @@ const TableToolbarAdvancedFilterFlyout = ({
...
+                    if (column.isDate) {
+                      return (
+                        <FormItem
+                          key={`simple-filter-item-${rowIndex}-${columnIndex}`}
+                          className={`${iotPrefix}--filter-flyout__simple-field`}
+                        >
+                          <DatePicker
+                            key={`${columnIndex}-${columnFilterValue}`}
+                            datePickerType="single"
+                            dateFormat="Y-m-d"
```
Comment (davidicus) [inline packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx:323]: We should allow for a way to customize the `dateFormat` and add a `locale`. Perhaps we can add to the `options` prop type to include an object for settings.

**Review by davidicus — CHANGES_REQUESTED** [review]
Hi @mapv04, thanks for the contribution! This is pretty good but there are a few items that need to be addressed. In addition to the changes listed in comments we will need to update `TableColumnsPropTypes` in `TablePropTypes.js`

File: packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx [hunk 2]
```diff
@@ -309,7 +311,39 @@ const TableToolbarAdvancedFilterFlyout = ({
...
+                                    ...prev.simple,
+                                    [column.id]: evt[0],
+                                  },
+                                };
+                              });
+                            }}
+                          >
+                            <DatePickerInput
+                              className={`${iotPrefix}--lelele`}
+                              placeholder="yyyy-mm-dd"
```
[the remark this answers is not in the document]
Comment (mapv04) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: I added the option to use a custom placeholder, also I updated the `TableColumnsPropTypes`

File: packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx (hunk 1 above)
Comment (mapv04) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: As the `options` prop is specific for the "select" fields, I created a dateOptions object to customize the `dateFormat` and `locale`.

Comment (mapv04) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: As the options prop is specific for the "select" fields, I created a dateOptions object to customize the dateFormat and locale.

File: packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidicus) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: Hey @mapv04. These changes look good but I am not seeing the `TablePropTypes.js` file included in your changes.

`packages/react/src/components/Table/TablePropTypes.js`

[the remark this answers is not in the document]
Comment (davidicus) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: That is the last items needed on our end to get this merged.

[the remark this answers is not in the document]
Comment (mapv04) [reply packages/react/src/components/Table/TableToolbar/TableToolbarAdvancedFilterFlyout.jsx]: Sorry I forgot to add the changes to the commit, done.

**Review by davidicus — APPROVED** [review]
This is great. Thanks again for the contribution. 

We have been talking internally about how to deal with these modifications in the future and we will probably implement a render function that can be used to render custom inputs. 

Something to keep in mind if you find you need more modifications to the current filters in the future.

**Review by jessieyan — APPROVED** [review]
Looks good to me

**Pull request merged**